[CWF] UESIM: Adding additional parameters to uesim.yml like msisdn, apn, rat-type …

sourabh-nanoti opened the pull request:
…while creating a ue in uesim

Signed-off-by: Sourabh Shekhar Nanoti <sourabhn@fb.com>

<!--
    Tag your PR title with the components that it touches.
    E.g. "[lte][agw] Changeset" or "[orc8r][docker] ..."
-->

## Summary
With this change, additional parameters like MSISDN, RAT-Type and APN can be configured for a UE. These parameters will be picked up from uesim.yml file. 
These additional parameters will help uesim when it tries to create a session in sessiond bypassing the aaa. 
## Test Plan

Tested this via uesim_cli. Please see the logs below:

vagrant@cwag-dev:~/magma/cwf/gateway/docker$ docker-compose exec uesim /bin/bash
WARNING: The ANALYTICS_GRAPHQL_TOKEN variable is not set. Defaulting to a blank string.
WARNING: The ANALYTICS_GRAPHQL_ENDPOINT variable is not set. Defaulting to a blank string.
root@cwag-dev:/# cd /var/opt/magma/bin/
root@cwag-dev:/var/opt/magma/bin# ./uesim_cli add_ue -auth_key 8BAF473F2F8FD09487CCCBD7097C6862 001010000000081  
I0812 05:19:31.182573      22 service_config.go:169] Successfully loaded '::uesim' service configs from '/etc/magma/uesim.yml'
I0812 05:19:31.182681      22 main.go:316] RAT-Type not set for Imsi[001010000000081], setting default rat 'NON_3GPP_EPS'
I0812 05:19:31.182689      22 main.go:322] Creating UE with IMSI:[001010000000081] MSISDN[9890971088] APN[wifi] RAT[6] 
Successfully added: 001010000000081
root@cwag-dev:/var/opt/magma/bin# ./uesim_cli add_ue -auth_key 8BAF473F2F8FD09487CCCBD7097C6862 001010000000082  
I0812 05:19:36.688651      32 service_config.go:169] Successfully loaded '::uesim' service configs from '/etc/magma/uesim.yml'
I0812 05:19:36.688735      32 main.go:316] RAT-Type not set for Imsi[001010000000082], setting default rat 'NON_3GPP_EPS'
I0812 05:19:36.688741      32 main.go:322] Creating UE with IMSI:[001010000000082] MSISDN[9890971089] APN[test] RAT[6] 
Successfully added: 001010000000082
root@cwag-dev:/var/opt/magma/bin# 

## Additional Information

sourabh-nanoti committed:
Adding additional parameters to uesim.yml like msisdn, apn, rat-type while creating a ue in uesim

Signed-off-by: Sourabh Shekhar Nanoti <sourabhn@fb.com>

sourabh-nanoti commented on cwf/gateway/tools/uesim_cli/main.go:
```diff
@@ -296,12 +302,37 @@ func createUeConfig(imsi string, auth_key []byte, seq_num uint64) (*protos.UECon
+			}
+			apn, ok = ue["apn"].(string)
+			if !ok {
+				glog.Infof("APN not set for Imsi[%s], setting default apn 'test'", imsi)
```
DefaultAPN can be printed here instead of hardcoded 'test'

sourabh-nanoti commented on cwf/gateway/tools/uesim_cli/main.go:
```diff
@@ -296,12 +302,37 @@ func createUeConfig(imsi string, auth_key []byte, seq_num uint64) (*protos.UECon
+			}
+			rat, ok = ue["rat"].(uint32)
+			if !ok {
+				glog.Infof("RAT-Type not set for Imsi[%s], setting default rat 'NON_3GPP_EPS'", imsi)
```
DefaultRatType can be printed here instead of hardcoded 'NON_3GPP_EPS'

sourabh-nanoti commented on cwf/gateway/configs/templates/uesim.yml:
```diff
@@ -0,0 +1,9 @@
+radius_secret: "888888"
+amf: "0000"
+subscribers:
+        "001010000000081": {msisdn: "9890971088", apn: "wifi", rat: 6}
```
Change this to some random number

sourabh-nanoti commented on cwf/gateway/configs/templates/uesim.yml:
```diff
@@ -0,0 +1,9 @@
+amf: "0000"
+subscribers:
+        "001010000000081": {msisdn: "9890971088", apn: "wifi", rat: 6}
+        "001010000000082": {msisdn: "9890971089", apn: "test"}
```
Change this to some random number

sourabh-nanoti committed:
Fixing small nits identified while self review

Signed-off-by: Sourabh Shekhar Nanoti <sourabhn@fb.com>

uri200 commented on cwf/protos/ue_sim.proto:
```diff
@@ -34,6 +34,15 @@ message UEConfig {
     // Sequence Number (SEQ).
     uint64 seq = 4;
+
+    // MSISDN
```
lets put those parameters inside an object so we can differenciate them from the standard configuration. Maybe `HSSless_config` object

sourabh-nanoti replied:
I think this proto files encapsulates the ue information which should be the same for with/without hss. Hence i added the parameters in the same object which will make the file more readable.

uri200 reviewed (changes requested):
We shouldn't start modifying the outer client, but the service itself. If you start from the outside you will impose conditions to the code inside the server and we can end up with messy code. So, if possible, always start from the inside out. 

In this case, since you need a new configuration, but we don't want to disturb the current code we should add this to the new `UESimServerHSSless` type

```

type UESimServerHSSless struct {
	UESimServer
}

func (srv *UESimServerHSSless) AddUE(ctx context.Context, ue *cwfprotos.UEConfig) (ret *protos.Void, err error) {

	// uses AddUE from the inner UESimServer to add UEs as normal
	ret, err = srv.UESimServer.AddUE(ctx, ue)
	if err != nil {
		return
	}

	// grab the UE you just added so you can include the new parameters
	ue, err = getUE(srv.store, ue.GetImsi())
	if err != nil {
		return
	}

	// ned to validate the data from HSSless
	err = validateUEDataHSSless(ue)
	if err != nil {
		err = ConvertStorageErrorToGrpcStatus(err)
		return
	}


	// The code comes from UESimServer.AddUE and stores again the UE.
	// This hould be written as a function this way it could be used by both
	// UESimServer and UESimServerHSSless

	blob, err := ueToBlob(ue)
	store, err := srv.store.StartTransaction(nil)
	if err != nil {
		err = errors.Wrap(err, "Error while starting transaction")
		err = ConvertStorageErrorToGrpcStatus(err)
		return
	}
	defer func() {
		switch err {
		case nil:
			if commitErr := store.Commit(); commitErr != nil {
				err = errors.Wrap(err, "Error while committing transaction")
				err = ConvertStorageErrorToGrpcStatus(err)
			}
		default:
			if rollbackErr := store.Rollback(); rollbackErr != nil {
				err = errors.Wrap(err, "Error while rolling back transaction")
				err = ConvertStorageErrorToGrpcStatus(err)
			}
		}
	}()

	err = store.CreateOrUpdate(networkIDPlaceholder, []blobstore.Blob{blob})
	return


```

uri200 commented on cwf/gateway/tools/uesim_cli/main.go:
```diff
@@ -296,12 +302,37 @@ func createUeConfig(imsi string, auth_key []byte, seq_num uint64) (*protos.UECon
 	if err != nil {
 		return nil, fmt.Errorf("could not generate OPc for subscriber: %s: %s", imsi, err)
 	}
+	ues, err := uecfg.GetMap("subscribers")
```
line 243 already grabs something called `subscribers`. Not sure if this can be used again

sourabh-nanoti replied:
We could refactor trafficgen and add ue commands to use single functions to read uedata from yml file.

sourabh-nanoti replied:
I will do this refactoring and post a new diff for this.

uri200 commented on cwf/gateway/tools/uesim_cli/main.go:
```diff
@@ -296,12 +302,37 @@ func createUeConfig(imsi string, auth_key []byte, seq_num uint64) (*protos.UECon
+	if err == nil {
+		ue, res := ues[imsi].(map[interface{}]interface{})
+		if res {
+			msisdn, ok = ue["msisdn"].(string)
```
all those parameters should not be grab here. Similar as imsi and auth key, they should be passed to this function. Or move all the UE parameters here. Otherwise we have the logic all scatter

sourabh-nanoti replied:
UE can be created in uesim from 2 different CLI commands add_ue and handleTrafficGenCmd so the logic is already scattered. We can refactor by making changes in the existing code with a risk of modifying stuff that is working.

sourabh-nanoti replied:
Posting a new diff with some refactor to make  the logic more clean.

uri200 commented on cwf/gateway/configs/templates/uesim.yml:
```diff
@@ -0,0 +1,9 @@
+radius_secret: "888888"
+amf: "0000"
+subscribers:
+        "001010000000081": {msisdn: "1234567890", apn: "wifi", rat: 6}
```
per `getConfiguredSubscribers` many parameters like `auth_key` are required, or the client will fail.  You will have to add fake values. Do not make them optional unless you add some logic to decide what parameters are needed for each case (either UEsim or UESimHSSless)

sourabh-nanoti replied:
I plan not make any parameters mandatory. In the add_ue command, OP is mandatory. What i feel is may be that can also be removed as it can be defaulted to the same OP for multiple IMSIs. But i did not made that change yet as its an existing code.

sourabh-nanoti replied:
I will address this in the refactoring diff that i am posting as an update to this diff.

sourabh-nanoti committed:
Refactoring duplicate for for reading UEs from yml file and incorporating code review comments.
Changes done in this commit

1. Removed OP as a mandatory parameter in the add_ue command while running uesim_cli. The OP now has to be passed via uesim.yml file
   using 'auth_key' as a key. Its cumbursome to add OP from CLI.
2. getConfiguredSubscribers() function now just gets the subscribers from the uesim.yml file. This API now takes 1 argument (imsi)
   if a single subscriber is to be added via add_ue command. Else this function will return all the UEs configured in the uesim.yml file.
3. Added a new function addUe() which will send the proto towards uesim for creating the UE.
4. Refactored the code for creating the ue proto from the yml file in the createUeConfig() function.
5. Fixed a segv in the handleTrafficGenCmd() regarding access of null WaitGroup variable 'wg'

Tests Done
1. Ran successful integ_test (although these changes do not change anything in that path)
2. Tried calling the handleTrafficGenCmd() and ensured that we dont get a segv
3. Added single IMSI via add_ue command and ensured that the ue is added in uesim along with custom and default parameters for rat, apn and msisdn.

Signed-off-by: Sourabh Shekhar Nanoti <sourabhn@fb.com>

sourabh-nanoti committed:
Refactoring duplicate for for reading UEs from yml file and incorporating code review comments.

Signed-off-by: Sourabh Shekhar Nanoti <sourabhn@fb.com>

sourabh-nanoti committed:
Merge branch 'UatAutomation' of github.com:sourabh-nanoti/magma into UatAutomation

uri200 commented:
Can we close this or resolve the conflicts?

sourabh-nanoti commented:
This can be closed. All the changes are  already  added as part of pull req #2917
